### PR TITLE
Census

### PR DESCRIPTION
**Pull request by Dominko**
The census is based on the application form, it operates based on a "census session" with custom questions which can change between years.
Every response generates a bunch of question response objects which can be aggregated and shown either on a specific page or exported to CSV.
The display page is a little rudimentary right now because I didn't get clear requirements and we can work on that once we have responses.
Currently only requirements to view/submit responses is being an active member. Since data is annonymised participation thought this was sufficient. Furthermore answering the census sets a cookie with the id of the census

### CODE REVIEW

File: census/models.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    class Meta(OrderedModel.Meta):
+        pass
+
+    session = models.ForeignKey(CensusSession, on_delete=CASCADE)
+    visible_in_voting = models.BooleanField(verbose_name='Visible for voters', default=True)
+    question_text = models.TextField(verbose_name='Question text', max_length=5000)
+    question_type = models.CharField(verbose_name='Type', max_length=20,
+                                     choices=[(tag.name, tag.value) for tag in QuestionType],
+                                     default=QuestionType.LongText)
+    question_options = models.TextField(verbose_name='Question options (optional)', max_length=300, blank=True)
```
Comment (bjin96) [inline census/models.py:69]: Making this only appear/writeable for choice questions could avoid potential confusion. A tool tip/note telling how to make options (I found out quite quickly newlines separate options, but can imagine confusion) might be helpful.

**Review by bjin96 — APPROVED** [review]
Thanks for all the work, Dominik! The census seems to work fine for me locally and overall looks good to me.

**An important comment regarding shuffling for anonymisation**, I noticed the shuffling happens just before display/export meaning the saved data would not be in random order. Pragmatically, I think that's fine, but technically this could leave room for finding individual answers from the stored data. If you think this might easily be changed so the answers are shuffled upon storing I think that might be preferrable.

I have also left a few comments in the code, mostly to help keeping the code base cleaner, please do have a look. I guess most aren't super important, but from my experience it's always better to add cleaned up code because cleaning it up later on will likely just not happen and keep piling up due to urgent feature merges until it's become a mess.

**Comment by bjin96** [comment]
Sorry, I've just had another thought regarding anonymity -- it would probably be good to prevent display/export of results when there's less than 10 responses to avoid the ability to reconstruct individual results :)

File: census/urls.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+from django.urls import path
+from . import views
+from django.contrib.auth.decorators import login_required
+from users.decorators import current_member_required, check_role
+
+app_name = 'census'
+
+urlpatterns = [
+    path('census/<int:session_id>/', login_required(current_member_required(views.CensusView.as_view())), name='census-form'),
```
[the remark this answers is not in the document]
Comment (Dominko) [reply census/urls.py:9]: Basically both achieve the same thing but I slightly prefer putting it in the URL as that's what django forums seemed to indicate is the "preferred" place of doing it

File: users/decorators.py [hunk 3]
```diff
@@ -26,18 +29,55 @@ def wrap(request, *args, **kwargs):
...
     return wrap
 
+def check_role(function, role_name):
+    def wrap(request, *args, **kwargs):
+        if Role.objects.filter(assigned_to=request.user, role_name=role_name).exists():
+            return function(request, *args, **kwargs)
+        else:
+            raise PermissionDenied
 
-def check_group(group_name):
```
[the remark this answers is not in the document]
Comment (Dominko) [reply users/decorators.py:30]: Its not used anywhere and also broken

**Pull request merged**